src/shared/switch-root.c

```diff
@@ -114,6 +114,14 @@ int switch_root(const char *new_root,
         if (r < 0) {
                 log_debug_errno(r, "Pivoting root file system failed, moving mounts instead: %m");
 
+                /* If we have to use MS_MOVE let's first try to get rid of *all* mounts we can, with the
+                 * exception of the path we want to switch to, plus everything leading to it and within
+                 * it. This is necessary because unlike pivot_root() just moving the mount to the root via
+                 * MS_MOVE won't magically unmount anything below it. Once the chroot() succeeds the mounts
+                 * below would still be around but invisible to us, because not accessible via
+                 * /proc/self/mountinfo. Hence, let's clean everything up first, as long as we still can. */
+                (void) umount_recursive_full(NULL, MNT_DETACH, STRV_MAKE(new_root));
```

Comment:
Note, `umount_recursive_full()` logs many lines in LOG_DEBUG.

Reply:
so these are MNT_DETACH umounts anyway (i.e. asynchronous umounts that do not wait for the superblock to be sync'ed), hence writable block devices might remain pinned elsewhere and we won't notice. MNT_DETACH is the only reliable thing to do here, since almost certainly various of the fs we deatch here might be pinned still. And MNT_DETACH typically doesn't fail btw, certainly not because of EBUSY or so. It's might fail due to MAC issues or so, but the whole point of it is that it always works, basically.

Also note that pivot_root() behaves the same (which is the syscall we'd rather use except that we cannot because kernel doesn't allow the original rootfs to be entirely detached, currently).

And yes, @yuwata is right, there's debug logging already in place.

i do agree that the fact that there might be file systems that suddenly become detached and invisible without synchronization guarantees sucks a bit. Hence I guess it would be wise to sync() before we do all this, so that even while there might later be hidden IO a correctly written file system wouldn't do this after sync8), and correct synchronization matters a lot more than hidden IO.

Reply:
→ #27700